fastcrypto-zkp/src/bn254/poseidon/mod.rs

```diff
@@ -70,29 +75,54 @@ pub fn hash(inputs: Vec<Fr>) -> Result<Fr, FastCryptoError> {
 }
 
 /// Calculate the poseidon hash of the field element inputs. If the input length is <= 16, calculate
-/// H(inputs), if it is <= 32, calculate H(H(inputs[0..16]), H(inputs[16..])), otherwise return an
-/// error.
+/// H(inputs), otherwise chunk the inputs into groups of 16, hash them and input the results recursively.
 pub fn to_poseidon_hash(inputs: Vec<Fr>) -> Result<Fr, FastCryptoError> {
-    if inputs.len() <= 16 {
+    if inputs.len() <= MERKLE_TREE_DEGREE {
         hash(inputs)
-    } else if inputs.len() <= 32 {
-        let hash1 = hash(inputs[0..16].to_vec())?;
-        let hash2 = hash(inputs[16..].to_vec())?;
-        hash([hash1, hash2].to_vec())
     } else {
-        Err(FastCryptoError::GeneralError(format!(
-            "Yet to implement: Unable to hash a vector of length {}",
-            inputs.len()
-        )))
+        to_poseidon_hash(
+            inputs
+                .chunks(MERKLE_TREE_DEGREE)
+                .map(|chunk| hash(chunk.to_vec()))
+                .collect::<FastCryptoResult<Vec<_>>>()?,
+        )
+    }
+}
+
+/// Calculate the poseidon hash of an array of inputs. Each input is interpreted as a BN254 field
+/// element assuming a little-endian encoding. The field elements are then hashed using the poseidon
+/// hash function ([to_poseidon_hash]) and the result is serialized as a little-endian integer (32
+/// bytes).
+///
+/// If one of the inputs is in non-canonical form, e.g. it represents an integer greater than the
+/// field size or is longer than 32 bytes, an error is returned.
+pub fn hash_to_bytes(
+    inputs: &Vec<Vec<u8>>,
```

Comment:
```suggestion
    inputs: &[[u8]]
```

Reply:
Unfortunately, this doesn't work because [u8] is not sized.